File: ruby/common/lib/common/resources/host.rb
# Copyright 2015 VMware, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, without warranties or
# conditions of any kind, EITHER EXPRESS OR IMPLIED. See the License for the
# specific language governing permissions and limitations under the License.

module EsxCloud
  class Host
    attr_accessor :id, :state, :username, :password, :availability_zone, :usage_tags, :address, :metadata, :esx_version

    # @param [HostCreateSpec] spec
    # @return [Host]
    def self.create(deployment_id, spec)
      Config.client.create_host(deployment_id, spec.to_hash)
    end

    # @param [String] json
    # @return [Host]
    def self.create_from_json(json)
      create_from_hash(JSON.parse(json))
    end

    # @param [Hash] hash
    # @return [Host]
    def self.create_from_hash(hash)
      unless hash.is_a?(Hash) && hash.keys.to_set.superset?(%w(id username password usageTags address metadata state)
        .to_set)
        raise UnexpectedFormat, "Invalid host hash: #{hash}"
      end
      new(hash["id"], hash["state"], hash["username"], hash["password"], hash["availabilityZone"], hash["usageTags"],
          hash["address"], hash["metadata"], hash["esxVersion"])
    end

    # @param [String] host_id
    # @return [Host]
    def self.find_host_by_id(host_id)
      Config.client.mgmt_find_host_by_id(host_id)
    end

    # @param [String] host_id
    # @return [Boolean]
    def self.delete(host_id)
      Config.client.mgmt_delete_host(host_id)
    end

    # @param [String] host_id
    # @return [VmList]
    def self.get_host_vms(host_id)
      Config.client.mgmt_get_host_vms(host_id)
    end

    # @param [String] host_id
    # @return [TaskList]
    def self.find_tasks_by_host_id(host_id)
      Config.client.find_tasks_by_host_id(host_id)
    end

    # @param [String] host_id
    # @return [Host]
    def self.enter_maintenance_mode(host_id)
      Config.client.host_enter_maintenance_mode(host_id)
    end

    # @param [String] host_id
    # @return [Host]
    def self.enter_suspended_mode(host_id)
      Config.client.host_enter_suspended_mode(host_id)
    end

    # @param [String] host_id
    # @return [Host]
    def self.exit_maintenance_mode(host_id)
      Config.client.host_exit_maintenance_mode(host_id)
    end

    # @param [String] host_id
    # @return [Host]
    def self.resume(host_id)
      Config.client.host_resume(host_id)
    end

    # @param [String] host_id
    # @param [HostSetAvailabilityZoneSpec] spec
    # @return [Host]
    def self.host_set_availability_zone(host_id, spec)
      Config.client.host_set_availability_zone(host_id, spec.to_hash)
    end

    # @param [String] id
    # @param [String] username
    # @param [String] password
    # @param [String] availability_zone
    # @param [Array<String>] usage_tags
    # @param [String] address
    # @param [Hash] metadata
    def initialize(id, state, username, password, availability_zone, usage_tags, address, metadata, esx_version)
      @id = id
      @state = state
      @username = username
      @password = password
      @availability_zone = availability_zone
      @usage_tags = usage_tags
      @address = address
      @metadata = metadata
      @esx_version = esx_version
    end

    def ==(other)
      @id == other.id && @state == other.state && @address == other.address && @username == other.username &&
          @password == other.password && @availability_zone == other.availability_zone &&
          @usage_tags == other.usage_tags && @metadata == other.metadata && @esx_version == other.esx_version
    end

    def to_spec()
      EsxCloud::HostCreateSpec.new(
        @username,
        @password,
        @usage_tags,
        @address,
        @metadata,
        @availability_zone)
    end
  end
end

File: java/api/src/main/java/com/vmware/photon/controller/api/Host.java
/*
 * Copyright 2015 VMware, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, without warranties or
 * conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.vmware.photon.controller.api;

import com.vmware.photon.controller.api.base.Base;
import com.vmware.photon.controller.api.constraints.DomainOrIP;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.wordnik.swagger.annotations.ApiModel;
import com.wordnik.swagger.annotations.ApiModelProperty;
import org.slf4j.LoggerFactory;

import javax.persistence.ElementCollection;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Host information returned from API.
 */
@ApiModel(value = "This class is the full fidelity representation of a Host object.")
@JsonIgnoreProperties(ignoreUnknown = true)
public class Host extends Base {
  public static final String KIND = "host";

  @JsonProperty
  @ApiModelProperty(value = "kind=\"host\"", required = true)
  private String kind = KIND;

  @JsonProperty
  @ApiModelProperty(value = "IP Address of Host", required = true)
  @DomainOrIP
  private String address;

  @JsonProperty
  @ApiModelProperty(value = "Host root username (typically \'root\')", required = true)
  @NotNull
  @Size(min = 1)
  private String username;

  @JsonProperty
  @ApiModelProperty(value = "Host password", required = true)
  @NotNull
  @Size(min = 1)
  private String password;

  @JsonProperty
  @ApiModelProperty(value = "Availability zone", required = false)
  private String availabilityZone;

  @ElementCollection
  @JsonProperty
  @ApiModelProperty(value = "Custom metadata that needs to be stored but not represented by individually " +
      "defined properties")
  private Map<String, String> metadata = new HashMap<String, String>();


  @JsonProperty
  @ApiModelProperty(value = "Usage tags", allowableValues = UsageTag.HOST_USAGES, required = true)
  @NotNull
  @Size(min = 1)
  private List<UsageTag> usageTags;

  @JsonProperty
  @ApiModelProperty(value = "ESX Version", required = false)
  private String esxVersion;

  @JsonProperty
  @ApiModelProperty(value = "Supplies the state of the Host",
      allowableValues = "CREATING,NOT_PROVISIONED,READY,MAINTENANCE,SUSPENDED,ERROR,DELETED.",
      required = true)
  private HostState state;

  @JsonIgnore
  private List<HostDatastore> datastores;

  public Host() {
  }

  public Host(String address, String username, String password, String availabilityZone, String esxVersion,
              List<UsageTag> usageTags, Map<String, String> metadata) {
    this.address = address;
    this.username = username;
    this.password = password;
    this.availabilityZone = availabilityZone;
    this.esxVersion = esxVersion;
    this.metadata.putAll(metadata);
    this.usageTags = usageTags;
  }

  @Override
  public String getKind() {
    return kind;
  }

  public String getAddress() {
    return address;
  }

  public void setAddress(String address) {
    this.address = address;
  }

  public String getUsername() {
    return username;
  }

  public void setUsername(String username) {
    this.username = username;
  }

  public String getPassword() {
    return password;
  }

  public void setPassword(String password) {
    this.password = password;
  }

  public String getAvailabilityZone() {
    return availabilityZone;
  }

  public void setAvailabilityZone(String availabilityZone) {
    this.availabilityZone = availabilityZone;
  }

  public List<UsageTag> getUsageTags() {
    return usageTags;
  }

  public void setUsageTags(List<UsageTag> usageTags) {
    this.usageTags = usageTags;
  }

  public String getEsxVersion() {
    return esxVersion;
  }

  public void setEsxVersion(String esxVersion) {
    this.esxVersion = esxVersion;
  }

  public Map<String, String> getMetadata() {
    return metadata;
  }

  public void setMetadata(Map<String, String> metadata) {
    this.metadata = metadata;
  }

  public HostState getState() {
    return state;
  }

  public void setState(HostState state) {
    this.state = state;
  }

  public List<HostDatastore> getDatastores() {
    return this.datastores;
  }

  public void setDatastores(List<HostDatastore> datastores) {
    LoggerFactory.getLogger(Host.class).debug("setDatastores {}", datastores);
    for (StackTraceElement ste : Thread.currentThread().getStackTrace()) {
      LoggerFactory.getLogger(Host.class).debug(ste.getFileName() + ste.getLineNumber());
    }
    this.datastores = datastores;
  }

  @Override
  protected com.google.common.base.Objects.ToStringHelper toStringHelper() {
    return super.toStringHelper()
        .add("address", address)
        .add("username", username)
        .add("password", password)
        .add("availabilityZone", availabilityZone)
        .add("esxVersion", esxVersion)
        .add("usageTags", usageTags)
        .add("metadata", metadata)
        .add("state", state)
        .add("datastores", datastores);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Host)) {
      return false;
    }
    if (!super.equals(o)) {
      return false;
    }

    Host that = (Host) o;
    return Objects.equals(username, that.username)
        && Objects.equals(password, that.password)
        && Objects.equals(address, that.address)
        && Objects.equals(availabilityZone, that.availabilityZone)
        && Objects.equals(esxVersion, that.esxVersion)
        && Objects.equals(usageTags, that.usageTags)
        && Objects.equals(metadata, that.metadata)
        && Objects.equals(datastores, that.datastores);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        super.hashCode(),
        address,
        username,
        password,
        availabilityZone,
        esxVersion,
        usageTags,
        metadata,
        datastores);
  }
}
